From 225b2d6ee0a4a7579629c8861440297119fd202f Mon Sep 17 00:00:00 2001
From: Ripan Deuri <quic_rdeuri@quicinc.com>
Date: Fri, 21 Mar 2025 14:54:11 +0530
Subject: [PATCH 5/8] wifi: ath12k: Hardware specific DP interrupt regisration

With the Unified Driver supporting multiple chipsets of different
hardware architectures, the interrupt handler may vary across hardware
architectures.

This patch introduces an architecture-specific ops table in ath12k_dp to
invoke functions defined in the architecture-specific module from the
common module. Use this ops table to register and unregister interrupt
handlers based on the underlying hardware architecture.

Tested-on: QCN9274 hw2.0 PCI WLAN.WBE.1.3.1-00173-QCAHKSWPL_SILICONZ-1
Tested-on: WCN7850 hw2.0 PCI WLAN.HMT.1.0.c5-00481-QCAHMTSWPL_V1.0_V2.0_SILICONZ-3

Signed-off-by: Ripan Deuri <quic_rdeuri@quicinc.com>
Signed-off-by: Harsh Kumar Bijlani <quic_hbijlani@quicinc.com>
Signed-off-by: Aman Mehta <quic_amanmeht@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h     |  3 +++
 drivers/net/wireless/ath/ath12k/dp.c       |  5 ++++
 drivers/net/wireless/ath/ath12k/dp.h       | 11 ++++++++
 drivers/net/wireless/ath/ath12k/hif.h      | 27 +++++++++++++++++++
 drivers/net/wireless/ath/ath12k/pci.c      |  2 ++
 drivers/net/wireless/ath/ath12k/pcic.c     | 20 +++++++--------
 drivers/net/wireless/ath/ath12k/pcic.h     |  6 +++++
 drivers/net/wireless/ath/ath12k/wifi7/dp.c | 30 +++++++++++++++++++---
 drivers/net/wireless/ath/ath12k/wifi7/dp.h |  2 --
 9 files changed, 90 insertions(+), 16 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 427fa6b..a24ee7b 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -167,6 +167,7 @@ enum ath12k_firmware_mode {
 #define ATH12K_MAX_TCL_RING_NUM	3
 
 struct ath12k_ext_irq_grp {
+	struct ath12k_dp *dp;
 	struct ath12k_base *ab;
 	u32 irqs[ATH12K_EXT_IRQ_NUM_MAX];
 	u32 num_irq;
@@ -179,6 +180,8 @@ struct ath12k_ext_irq_grp {
 #else
 	struct net_device napi_ndev;
 #endif
+	int (*irq_handler)(struct ath12k_dp *dp,
+			   struct ath12k_ext_irq_grp *irq_grp, int budget);
 };
 
 enum ath12k_smbios_cc_type {
diff --git a/drivers/net/wireless/ath/ath12k/dp.c b/drivers/net/wireless/ath/ath12k/dp.c
index 50effeb..541fe15 100644
--- a/drivers/net/wireless/ath/ath12k/dp.c
+++ b/drivers/net/wireless/ath/ath12k/dp.c
@@ -1633,6 +1633,7 @@ fail_link_desc_cleanup:
 
 void ath12k_dp_cmn_device_deinit(struct ath12k_dp *dp)
 {
+	dp->arch_ops->dp_op_device_deinit(dp);
 	ath12k_dp_cleanup(dp->ab);
 }
 
@@ -1644,6 +1645,10 @@ int ath12k_dp_cmn_device_init(struct ath12k_dp *dp)
 	if (ret)
 		return ret;
 
+	ret = dp->arch_ops->dp_op_device_init(dp);
+	if (ret)
+		return ret;
+
 	return 0;
 }
 
diff --git a/drivers/net/wireless/ath/ath12k/dp.h b/drivers/net/wireless/ath/ath12k/dp.h
index f2e84cb..5d33305 100644
--- a/drivers/net/wireless/ath/ath12k/dp.h
+++ b/drivers/net/wireless/ath/ath12k/dp.h
@@ -341,6 +341,15 @@ struct ath12k_link_stats {
 	u32 tx_desc_type[HAL_TCL_DESC_TYPE_MAX];
 };
 
+/* DP arch ops to communicate from common module
+ * to arch specific module
+ */
+struct ath12k_dp_arch_ops {
+	int (*dp_op_device_init)(struct ath12k_dp *dp);
+	void (*dp_op_device_deinit)(struct ath12k_dp *dp);
+};
+
+
 struct ath12k_dp {
 	struct ath12k_base *ab;
 	u8 num_bank_profiles;
@@ -401,6 +410,8 @@ struct ath12k_dp {
 	struct device *dev;
 	struct ath12k_dp_hw_group *dp_hw_grp;
 	u8 device_id;
+
+	struct ath12k_dp_arch_ops *arch_ops;
 };
 
 static inline void ath12k_dp_get_mac_addr(u32 addr_l32, u16 addr_h16, u8 *addr)
diff --git a/drivers/net/wireless/ath/ath12k/hif.h b/drivers/net/wireless/ath/ath12k/hif.h
index 24bf5bb..0af243d 100644
--- a/drivers/net/wireless/ath/ath12k/hif.h
+++ b/drivers/net/wireless/ath/ath12k/hif.h
@@ -34,6 +34,12 @@ struct ath12k_hif_ops {
 	void (*get_ce_msi_idx)(struct ath12k_base *ab, u32 ce_id, u32 *msi_idx);
 	int (*panic_handler)(struct ath12k_base *ab);
 	void (*coredump_download)(struct ath12k_base *ab);
+	int (*ext_irq_setup)(struct ath12k_base *ab,
+			     int (*handler)(struct ath12k_dp *dp,
+					    struct ath12k_ext_irq_grp *irq_grp,
+					    int budget),
+			     struct ath12k_dp *dp);
+	void (*ext_irq_cleanup)(struct ath12k_base *ab);
 	int (*set_qrtr_endpoint_id)(struct ath12k_base *ab);
 	void (*config_static_window)(struct ath12k_base *ab);
 	int (*get_msi_irq)(struct ath12k_base *ab, unsigned int vector);
@@ -198,4 +204,25 @@ static inline int ath12k_hif_get_msi_irq(struct ath12k_base *ab, unsigned int ve
 	return ab->hif.ops->get_msi_irq(ab, vector);
 }
 
+static inline
+int ath12k_hif_ext_irq_setup(struct ath12k_base *ab,
+			     int (*irq_handler)(struct ath12k_dp *dp,
+						struct ath12k_ext_irq_grp *irq_grp,
+						int budget),
+			     struct ath12k_dp *dp)
+{
+	if (!ab->hif.ops->ext_irq_setup)
+		return -EOPNOTSUPP;
+
+	return ab->hif.ops->ext_irq_setup(ab, irq_handler, dp);
+}
+
+static inline void ath12k_hif_ext_irq_cleanup(struct ath12k_base *ab)
+{
+	if (!ab->hif.ops->ext_irq_cleanup)
+		return;
+
+	ab->hif.ops->ext_irq_cleanup(ab);
+}
+
 #endif /* ATH12K_HIF_H */
diff --git a/drivers/net/wireless/ath/ath12k/pci.c b/drivers/net/wireless/ath/ath12k/pci.c
index 0c7dbcf..1506e00 100644
--- a/drivers/net/wireless/ath/ath12k/pci.c
+++ b/drivers/net/wireless/ath/ath12k/pci.c
@@ -1030,6 +1030,8 @@ static const struct ath12k_hif_ops ath12k_pci_hif_ops = {
 #ifdef CPTCFG_ATH12K_COREDUMP
 	.coredump_download = ath12k_pci_coredump_download,
 #endif
+	.ext_irq_setup = ath12k_pcic_ext_irq_config,
+	.ext_irq_cleanup = ath12k_pcic_free_ext_irq,
 };
 
 static enum ath12k_device_family
diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index 42dc2bb..3d8b778 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -174,7 +174,7 @@ u32 ath12k_pcic_get_window_start(struct ath12k_base *ab, u32 offset)
 	return window_start;
 }
 
-static void ath12k_pcic_free_ext_irq(struct ath12k_base *ab)
+void ath12k_pcic_free_ext_irq(struct ath12k_base *ab)
 {
 	int i, j;
 
@@ -201,8 +201,6 @@ void ath12k_pcic_free_irq(struct ath12k_base *ab)
 		irq_idx = ATH12K_PCI_IRQ_CE0_OFFSET + i;
 		free_irq(ab->irq_num[irq_idx], &ab->ce.ce_pipe[i]);
 	}
-
-	ath12k_pcic_free_ext_irq(ab);
 }
 
 void ath12k_pcic_free_hybrid_irq(struct ath12k_base *ab)
@@ -321,10 +319,9 @@ static int ath12k_pcic_ext_grp_napi_poll(struct napi_struct *napi, int budget)
 	struct ath12k_ext_irq_grp *irq_grp = container_of(napi,
 						struct ath12k_ext_irq_grp,
 						napi);
-	struct ath12k_base *ab = irq_grp->ab;
 	int work_done;
 
-	work_done = ath12k_wifi7_dp_service_srng(ab, irq_grp, budget);
+	work_done = irq_grp->irq_handler(irq_grp->dp, irq_grp, budget);
 	if (work_done < budget) {
 		napi_complete_done(napi, work_done);
 		ath12k_pcic_ext_grp_enable(irq_grp);
@@ -686,7 +683,11 @@ void ath12k_pcic_ipci_write32(struct ath12k_base *ab, u32 offset, u32 value)
 		  (offset & WINDOW_RANGE_MASK));
 }
 
-static int ath12k_pci_ext_irq_config(struct ath12k_base *ab)
+int ath12k_pcic_ext_irq_config(struct ath12k_base *ab,
+			       int (*irq_handler)(struct ath12k_dp *dp,
+						  struct ath12k_ext_irq_grp *irq_grp,
+						  int budget),
+			       struct ath12k_dp *dp)
 {
 
 	struct ath12k_pci *ar_pci = (struct ath12k_pci *)ab->drv_priv;
@@ -710,7 +711,8 @@ static int ath12k_pci_ext_irq_config(struct ath12k_base *ab)
 
 		irq_grp->ab = ab;
 		irq_grp->grp_id = i;
-
+		irq_grp->irq_handler = irq_handler;
+		irq_grp->dp = dp;
 #if LINUX_VERSION_IS_GEQ(6,10,0)
 		irq_grp->napi_ndev = alloc_netdev_dummy(0);
 		napi_ndev = irq_grp->napi_ndev;
@@ -840,10 +842,6 @@ int ath12k_pcic_config_irq(struct ath12k_base *ab)
 		ath12k_pcic_ce_irq_disable(ab, i);
 	}
 
-	ret = ath12k_pci_ext_irq_config(ab);
-	if (ret)
-		return ret;
-
 	return 0;
 }
 
diff --git a/drivers/net/wireless/ath/ath12k/pcic.h b/drivers/net/wireless/ath/ath12k/pcic.h
index cf1c428..467fe1f 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.h
+++ b/drivers/net/wireless/ath/ath12k/pcic.h
@@ -72,4 +72,10 @@ int ath12k_pcic_get_msi_irq(struct ath12k_base *ab, unsigned int vector);
 int ath12k_pcic_config_hybrid_irq(struct ath12k_base *ab);
 int ath12k_pcic_config_irq(struct ath12k_base *ab);
 void ath12k_pcic_free_irq(struct ath12k_base *ab);
+int ath12k_pcic_ext_irq_config(struct ath12k_base *ab,
+			       int (*irq_handler)(struct ath12k_dp *dp,
+						  struct ath12k_ext_irq_grp *irq_grp,
+						  int budget),
+			       struct ath12k_dp *dp);
+void ath12k_pcic_free_ext_irq(struct ath12k_base *ab);
 #endif
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp.c b/drivers/net/wireless/ath/ath12k/wifi7/dp.c
index 1a59098..00b364d 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp.c
@@ -7,6 +7,7 @@
 #include "../debug.h"
 #include "../dp_rx.h"
 #include "../dp_tx.h"
+#include "../hif.h"
 #include "../dp_mon.h"
 #include "../dp_cmn.h"
 #include "dp_rx.h"
@@ -14,10 +15,11 @@
 #include "dp_tx.h"
 #include "dp_rx.h"
 
-int ath12k_wifi7_dp_service_srng(struct ath12k_base *ab,
-				 struct ath12k_ext_irq_grp *irq_grp,
-				 int budget)
+static int ath12k_wifi7_dp_service_srng(struct ath12k_dp *dp,
+					struct ath12k_ext_irq_grp *irq_grp,
+					int budget)
 {
+	struct ath12k_base *ab = dp->ab;
 	struct napi_struct *napi = &irq_grp->napi;
 	int grp_id = irq_grp->grp_id;
 	int work_done = 0;
@@ -116,6 +118,26 @@ done:
 	return tot_work_done;
 }
 
+static int ath12k_wifi7_dp_op_device_init(struct ath12k_dp *dp)
+{
+	int ret;
+
+	ret = ath12k_hif_ext_irq_setup(dp->ab, ath12k_wifi7_dp_service_srng, dp);
+
+	return ret;
+}
+
+static void ath12k_wifi7_dp_op_device_deinit(struct ath12k_dp *dp)
+{
+	ath12k_hif_ext_irq_cleanup(dp->ab);
+}
+
+static struct ath12k_dp_arch_ops ath12k_wifi7_dp_arch_ops = {
+	.dp_op_device_init = ath12k_wifi7_dp_op_device_init,
+	.dp_op_device_deinit = ath12k_wifi7_dp_op_device_deinit,
+
+};
+
 /* TODO: remove export once this file is built with wifi7 ko */
 struct ath12k_dp *ath12k_wifi7_dp_init(struct ath12k_base *ab)
 {
@@ -125,6 +147,8 @@ struct ath12k_dp *ath12k_wifi7_dp_init(struct ath12k_base *ab)
 	if (!dp)
 		return NULL;
 
+	dp->arch_ops = &ath12k_wifi7_dp_arch_ops;
+
 	dp->ab = ab;
 	dp->dev = ab->dev;
 	dp->hw_params = ab->hw_params;
diff --git a/drivers/net/wireless/ath/ath12k/wifi7/dp.h b/drivers/net/wireless/ath/ath12k/wifi7/dp.h
index da70e1b..948cafb 100644
--- a/drivers/net/wireless/ath/ath12k/wifi7/dp.h
+++ b/drivers/net/wireless/ath/ath12k/wifi7/dp.h
@@ -13,8 +13,6 @@
 struct ath12k_base;
 struct ath12k_dp;
 
-int ath12k_wifi7_dp_service_srng(struct ath12k_base *ab,
-				 struct ath12k_ext_irq_grp *irq_grp, int budget);
 struct ath12k_dp *ath12k_wifi7_dp_init(struct ath12k_base *ab);
 void ath12k_wifi7_dp_deinit(struct ath12k_dp *dp);
 
-- 
2.34.1

